File: тесты/bench04_matrix_mul.py
import time

def mul(A, B):
    return [
        A[0] * B[0] + A[1] * B[2],
        A[0] * B[1] + A[1] * B[3],
        A[2] * B[0] + A[3] * B[2],
        A[2] * B[1] + A[3] * B[3]
    ]

def power(M, p):
    res = [1.0, 0.0, 0.0, 1.0]
    base = [M[0], M[1], M[2], M[3]]
    while p > 0:
        if p % 2 == 1:
            res = mul(res, base)
        base = mul(base, base)
        p //= 2
    return res

if __name__ == "__main__":
    t0 = time.perf_counter()
    ans = 0
    T = [1.0, 1.0, 1.0, 0.0]
    k = 0
    while k < 500000:
        res = power(T, 30)
        ans = res[0]
        k += 1
    t1 = time.perf_counter()
    ms = (t1 - t0) * 1000.0
    print(f"RESULT:{int(ans)}")
    print(f"TIME_MS:{ms:.2f}")

File: тесты/bench04_matrix_mul.cpp
#include <iostream>
#include <chrono>

struct Mat2 {
    double a, b, c, d;
};

Mat2 mul(const Mat2& X, const Mat2& Y) {
    return {
        X.a * Y.a + X.b * Y.c,
        X.a * Y.b + X.b * Y.d,
        X.c * Y.a + X.d * Y.c,
        X.c * Y.b + X.d * Y.d
    };
}

Mat2 power(Mat2 M, int p) {
    Mat2 res = {1.0, 0.0, 0.0, 1.0};
    Mat2 base = M;
    while (p > 0) {
        if (p % 2 == 1) res = mul(res, base);
        base = mul(base, base);
        p /= 2;
    }
    return res;
}

int main() {
    auto t0 = std::chrono::high_resolution_clock::now();
    double ans = 0;
    Mat2 T = {1.0, 1.0, 1.0, 0.0};
    for (int k = 0; k < 500000; k++) {
        Mat2 res = power(T, 30);
        ans = res.a;
    }
    auto t1 = std::chrono::high_resolution_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::cout << "RESULT:" << (long long)ans << std::endl;
    std::cout << "TIME_MS:" << ms << std::endl;
    return 0;
}
